### PR TITLE
Add installDirs (datadir, etc) to new-build and remove them from new-install

### PR DESCRIPTION
**Pull request by gbaz**
Resolves https://github.com/haskell/cabal/issues/3586

Handles the --datadir flag, etc. in new-build. Also adds that flag to those that can be parsed from cabal.project files.

Further, removes those flags, which never worked, from `cabal new-install`, as the new install command, by design, always places things in a fixed location in the store.

* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

### CODE REVIEW

**Review by Mikolaj — APPROVED** [review]
This seems a sensible thing to do.

**Comment by jneira** [comment]
nice fix, it would need a changelog, no?

**Comment by Mikolaj** [comment]
@gbaz: could you add a tiny changelog? thank you!

File: cabal-install/src/Distribution/Client/CmdInstall.hs [hunk 1]
```diff
@@ -168,9 +168,14 @@ installCommand = CommandUI
...
       ++ "    Install the package in the ./pkgfoo directory\n"
 
-  , commandOptions      = nixStyleOptions clientInstallOptions
+  , commandOptions      = \x -> filter notInstallDirOpt $ nixStyleOptions clientInstallOptions x
   , commandDefaultFlags = defaultNixStyleFlags defaultClientInstallFlags
   }
+ where
+  -- install doesn't take installDirs flags, since it always installs into the store in a fixed way.
+  notInstallDirOpt x = not $ optionName x `elem` installDirOptNames
+  installDirOptNames = map optionName installDirsOptions
```
Comment (Ericson2314) [inline cabal-install/src/Distribution/Client/CmdInstall.hs:177]: Is this how command line flags are usually filtered out? I would think we would have plenty of of cases of command `foo` handling some flags that are really just dispatched to the underlying command `bar`?

File: cabal-install/src/Distribution/Client/ProjectConfig/Types.hs [hunk 2]
```diff
@@ -172,7 +172,7 @@ data ProjectConfigShared
        -- Things that only make sense for manual mode, not --local mode
        -- too much control!
      --projectConfigUserInstall       :: Flag Bool,
-     --projectConfigInstallDirs       :: InstallDirs (Flag PathTemplate),
+       projectConfigInstallDirs       :: InstallDirs (Flag PathTemplate),
```
Comment (Ericson2314) [inline cabal-install/src/Distribution/Client/ProjectConfig/Types.hs:175]: Do we want to move this, or does the comment above still apply?

**Comment by Mikolaj** [comment]
@gbaz: we are waiting in #8520 to test if this solves their use case. Any chance to merge this soon?

File: cabal-install/src/Distribution/Client/CmdInstall.hs (hunk 1 above)
Comment (gbaz) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:177]: The issue is that nixStyleOptions _adds_ these things as options. So in this particular case we have nix style options which we specifically want to filter out from a particular v2 command.

**Comment by gbaz** [comment]
Added a changelog and consider this ready to merge. Just need a second review!

**Review by andreasabel — APPROVED** [review]

**Comment by andreasabel** [comment]
Needs **SQUASHING**.

**Comment by Mikolaj** [comment]
@gbaz: I've added the label in the interest of swiftness. Please feel free to change the label in the next 2 days, before it gets merged.

**Comment by Mikolaj** [comment]
Actually, @gbaz, I missed the merge conflict. :(

**Pull request merged**